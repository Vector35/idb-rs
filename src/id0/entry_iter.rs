use std::marker::PhantomData;

use anyhow::{anyhow, Result};

use crate::id0::{get_netnode_from_key, get_tag_from_key, is_key_netnode};
use crate::IDAKind;

use super::{get_sup_from_key, ID0Entry, ID0Section, NetnodeIdx};

pub struct EntryTagSubkeys<'a, K: IDAKind> {
    entries: &'a [ID0Entry],
    netnode: NetnodeIdx<K>,
    tag: u8,
}

impl<'a, K: IDAKind> EntryTagSubkeys<'a, K> {
    pub fn new(
        entries: &'a [ID0Entry],
        netnode: NetnodeIdx<K>,
        tag: u8,
    ) -> Self {
        Self {
            entries,
            netnode,
            tag,
        }
    }
}

impl<'a, K: IDAKind> Iterator for EntryTagSubkeys<'a, K> {
    type Item = &'a ID0Entry;

    fn next(&mut self) -> Option<Self::Item> {
        let (current, rest) = self.entries.split_first()?;
        if !is_key_netnode(&current.key)
            || get_netnode_from_key::<K>(&current.key) != Some(self.netnode.0)
            || get_tag_from_key::<K>(&current.key) != Some(self.tag)
        {
            return None;
        }
        self.entries = rest;
        Some(current)
    }
}

pub struct EntryTagContinuousSubkeys<'a, K: IDAKind> {
    iter: EntryTagSubkeys<'a, K>,
    expected_alt: K::Usize,
}

impl<'a, K: IDAKind> EntryTagContinuousSubkeys<'a, K> {
    pub fn new(
        id0: &'a ID0Section<K>,
        netnode: NetnodeIdx<K>,
        tag: u8,
        start: K::Usize,
    ) -> Self {
        let start_idx =
            id0.netnode_tag_alt_idx(netnode, start, tag).unwrap_or(0);
        let entries = &id0.entries[start_idx..];
        let expected_alt = entries
            .get(0)
            .and_then(|entry| get_sup_from_key::<K>(&entry.key))
            .unwrap_or(0u8.into());
        Self {
            iter: EntryTagSubkeys::new(entries, netnode, tag),
            expected_alt,
        }
    }
}

impl<'a, K: IDAKind> Iterator for EntryTagContinuousSubkeys<'a, K> {
    type Item = &'a ID0Entry;

    fn next(&mut self) -> Option<Self::Item> {
        let current = self.iter.next()?;
        if get_sup_from_key::<K>(&current.key) != Some(self.expected_alt) {
            return None;
        }
        self.expected_alt += 1u8.into();
        Some(current)
    }
}

#[derive(Clone, Copy)]
pub struct NetnodeRangeIter<'a, K> {
    pub(crate) entries: &'a [ID0Entry],
    _kind: PhantomData<K>,
}

impl<'a, K> NetnodeRangeIter<'a, K> {
    pub fn new(entries: &'a [ID0Entry]) -> Self {
        Self {
            entries,
            _kind: PhantomData,
        }
    }
}

impl<'a, K: IDAKind> Iterator for NetnodeRangeIter<'a, K> {
    type Item = (&'a [u8], &'a [u8]);

    fn next(&mut self) -> Option<Self::Item> {
        let key_len = super::key_len_netnode::<K>();
        let (current, rest) = self.entries.split_first()?;
        self.entries = rest;
        Some((&current.key[key_len..], &current.value[..]))
    }
}

#[derive(Debug, Clone, Copy)]
pub struct NetnodeSupRangeIter<'a, K> {
    pub(crate) entries: &'a [ID0Entry],
    _kind: PhantomData<K>,
}

impl<'a, K> NetnodeSupRangeIter<'a, K> {
    pub fn new(entries: &'a [ID0Entry]) -> Self {
        Self {
            entries,
            _kind: PhantomData,
        }
    }
}

impl<'a, K: IDAKind> Iterator for NetnodeSupRangeIter<'a, K> {
    type Item = Result<(K::Usize, &'a [u8])>;

    fn next(&mut self) -> Option<Self::Item> {
        let key_len = super::key_len_netnode_tag::<K>();
        let (current, rest) = self.entries.split_first()?;
        self.entries = rest;
        let key_raw = &current.key[key_len..];
        let Some(key) = K::usize_try_from_be_bytes(key_raw) else {
            return Some(Err(anyhow!("Invalid sup netnode key")));
        };
        Some(Ok((key, &current.value[..])))
    }
}
