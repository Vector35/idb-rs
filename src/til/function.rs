use crate::ida_reader::{IdaGenericBufUnpack, IdaGenericUnpack};
use crate::til::section::TILSectionHeader;
use crate::til::{associate_field_name_and_member, Basic, Type, TypeMetadata, TypeRaw, TAH};
use anyhow::{ensure, Context};

#[derive(Debug, Clone)]
pub struct Function {
    pub ret: Box<Type>,
    pub args: Vec<(Option<Vec<u8>>, Type, Option<ArgLoc>)>,
    pub retloc: Option<ArgLoc>,
}
impl Function {
    pub(crate) fn new(
        til: &TILSectionHeader,
        value: FunctionRaw,
        fields: Option<Vec<Vec<u8>>>,
    ) -> anyhow::Result<Self> {
        let args = associate_field_name_and_member(fields, value.args)
            .context("Function")?
            .map(|(n, (t, a))| Type::new(til, t, None).map(|t| (n, t, a)))
            .collect::<anyhow::Result<_, _>>()?;
        Ok(Self {
            ret: Type::new(til, *value.ret, None).map(Box::new)?,
            args,
            retloc: value.retloc,
        })
    }
}

#[derive(Debug, Clone)]
pub(crate) struct FunctionRaw {
    pub ret: Box<TypeRaw>,
    pub args: Vec<(TypeRaw, Option<ArgLoc>)>,
    pub retloc: Option<ArgLoc>,
}

#[derive(Debug, Clone)]
pub enum ArgLoc {
    // TODO add those to flags
    // ::ALOC_STACK
    // ::ALOC_STATIC
    // ::ALOC_REG1
    // ::ALOC_REG2
    // ::ALOC_RREL
    // ::ALOC_DIST
    // ::ALOC_CUSTOM
    /// 0 - None
    None,
    /// 1 - stack offset
    Stack(u32),
    /// 2 - distributed (scattered)
    Dist(Vec<ArgLocDist>),
    /// 3 - one register (and offset within it)
    Reg1(u32),
    /// 4 - register pair
    Reg2(u32),
    /// 5 - register relative
    RRel { reg: u16, off: u32 },
    /// 6 - global address
    Static(u32),
    // 7..=0xf custom
    // TODO is possible to know the custom impl len?
}

#[derive(Debug, Clone)]
pub struct ArgLocDist {
    pub info: u16,
    pub off: u16,
    pub size: u16,
}

impl FunctionRaw {
    pub(crate) fn read(
        input: &mut impl IdaGenericBufUnpack,
        header: &TILSectionHeader,
        metadata: u8,
    ) -> anyhow::Result<Self> {
        // TODO what is that?
        let mut flags = metadata << 2;

        let cc = Self::read_cc(&mut *input, &mut flags)?;

        let _tah = TAH::read(&mut *input)?;
        let ret = TypeRaw::read(&mut *input, header)?;
        // TODO double check documentation for [flag::tf_func::BT_FUN]
        let have_retloc = cc.get_calling_convention().is_special_pe()
            && !matches!(&ret, TypeRaw::Basic(Basic::Void));
        let retloc = have_retloc.then(|| ArgLoc::read(&mut *input)).transpose()?;
        if cc.get_calling_convention().is_void_arg() {
            return Ok(Self {
                ret: Box::new(ret),
                args: vec![],
                retloc,
            });
        }

        let n = input.read_dt()?;
        let is_special_pe = cc.get_calling_convention().is_special_pe();
        let args = (0..n)
            .map(|_| -> anyhow::Result<_> {
                let tmp = input.fill_buf()?.first().copied();
                if tmp == Some(0xFF) {
                    // TODO what is this?
                    let _tmp: u8 = bincode::deserialize_from(&mut *input)?;
                    let _flags = input.read_de()?;
                }
                let tinfo = TypeRaw::read(&mut *input, header)?;
                let argloc = is_special_pe
                    .then(|| ArgLoc::read(&mut *input))
                    .transpose()?;

                Ok((tinfo, argloc))
            })
            .collect::<anyhow::Result<_, _>>()?;

        Ok(Self {
            ret: Box::new(ret),
            args,
            retloc,
        })
    }

    /// [BT_FUNC](https://hex-rays.com/products/ida/support/sdkdoc/group__tf__func.html#ga7b7fee21f21237beb6d91e854410e0fa)
    fn read_cc(
        input: &mut impl IdaGenericBufUnpack,
        flags: &mut u8,
    ) -> anyhow::Result<TypeMetadata> {
        let mut cm = TypeMetadata::read(&mut *input)?;
        if !cm.get_calling_convention().is_spoiled() {
            return Ok(cm);
        }
        // TODO find what to do with this spoiled and flags stuff
        let mut _spoiled = vec![];
        loop {
            // TODO create flags::CM_CC_MASK
            let nspoiled = cm.0 & 0xf;
            if nspoiled == 0xF {
                let bfa_byte: u8 = bincode::deserialize_from(&mut *input)?;
                if bfa_byte & 0x80 != 0 {
                    // TODO what is this? Do this repeat `bfa_byte & 0xF` number of times?
                    let _fti_bits: u16 = bincode::deserialize_from(&mut *input)?;
                } else {
                    *flags |= (bfa_byte & 0x1F) << 1;
                }
            } else {
                for _ in 0..nspoiled {
                    let b: u8 = bincode::deserialize_from(&mut *input)?;
                    let (size, reg) = if b & 0x80 != 0 {
                        let size: u8 = bincode::deserialize_from(&mut *input)?;
                        let reg = b & 0x7F;
                        (size, reg)
                    } else {
                        ensure!(b > 0, "Unable to solve register from a spoiled function");
                        let size = (b >> 4) + 1;
                        let reg = (b & 0xF) - 1;
                        (size, reg)
                    };
                    _spoiled.push((size, reg));
                }
                *flags |= 1;
            }

            cm = TypeMetadata::read(&mut *input)?;
            if !cm.get_calling_convention().is_spoiled() {
                return Ok(cm);
            }
        }
    }
}

impl ArgLoc {
    fn read(input: &mut impl IdaGenericUnpack) -> anyhow::Result<Self> {
        let t: u8 = input.read_u8()?;
        if t != 0xFF {
            let b = t & 0x7F;
            match (t, b) {
                (0..=0x80, 1..) => Ok(Self::Reg1((b - 1).into())),
                (0..=0x80, 0) => Ok(Self::Stack(0)),
                _ => {
                    let c: u8 = bincode::deserialize_from(&mut *input)?;
                    if c == 0 {
                        Ok(Self::None)
                    } else {
                        Ok(Self::Reg2(u32::from(b) | u32::from(c - 1) << 16))
                    }
                }
            }
        } else {
            let typ = input.read_dt()?;
            match typ & 0xF {
                0 => Ok(Self::None),
                1 => {
                    let sval = input.read_de()?;
                    Ok(Self::Stack(sval))
                }
                2 => {
                    let n = (typ >> 5) & 0x7;
                    let dist: Vec<_> = (0..n)
                        .map(|_| {
                            let info = input.read_dt()?;
                            let off = input.read_dt()?;
                            let size = input.read_dt()?;
                            Ok(ArgLocDist { info, off, size })
                        })
                        .collect::<anyhow::Result<_>>()?;
                    Ok(Self::Dist(dist))
                }
                3 => {
                    let reg_info = input.read_dt()?;
                    // TODO read other dt?
                    Ok(Self::Reg1(reg_info.into()))
                }
                4 => {
                    let reg_info = input.read_dt()?;
                    // TODO read other dt?
                    Ok(Self::Reg2(reg_info.into()))
                }
                5 => {
                    let reg = input.read_dt()?;
                    let off = input.read_de()?;
                    Ok(Self::RRel { reg, off })
                }
                6 => {
                    let sval = input.read_de()?;
                    Ok(Self::Static(sval))
                }
                0x7..=0xF => todo!("Custom implementation for ArgLoc"),
                _ => unreachable!(),
            }
        }
    }
}
