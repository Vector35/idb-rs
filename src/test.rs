use rstest::rstest;

use std::fs::File;
use std::io::Cursor;
use std::path::{Path, PathBuf};

use crate::id0::function::FunctionsAndComments;
use crate::id0::{FileRegions, Segment};
use crate::*;

#[test]
fn parse_id0_til() {
    let function = [
        0x0c, // Function Type
        0xaf, 0x81, 0x42, 0x01, 0x53, // TODO
        0x01, // void ret
        0x03, //n args
        0x3d, 0x08, 0x48, 0x4d, 0x4f, 0x44, 0x55, 0x4c, 0x45, 0x3d, 0x06, 0x44,
        0x57, 0x4f, 0x52, 0x44, 0x00,
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_destructor_function() {
    // from ComRAT-Orchestrator.i64 0x180007cf0
    // ```c
    // void __fastcall stringstream__basic_ios__sub_180007CF0_Destructor(
    //   basic_ios *__shifted(stringstream,0x94) a1
    // );
    // ```
    let function = [
        0x0c, // Function Type
        0x70, // TODO
        0x01, // void ret
        0x02, // n args (2 - 1 = 1)
        0x0a, // arg0 type is pointer
        0xfe, 0x80, 0x01, // pointer tah
        0x3d, // pointer type is typedef
        0x04, 0x23, 0x83, 0x69, // typedef ord is 233 -> basic_ios
        // ?????
        0x3d, // second typedef?
        0x04, 0x23, 0x83, 0x66, // typedef ord is 230 => stringstream
        0x82, 0x54, // ???? the 0x94 value?
        0x00, // the final value always present
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_function_ext_att() {
    // ```
    // Function {
    //   ret: Basic(Int { bytes: 4, is_signed: None }),
    //   args: [(
    //     Some("env"),
    //     Pointer(Pointer {
    //       closure: Default,
    //       tah: TAH(TypeAttribute(1)),
    //       typ: Struct(Ref {
    //         ref_type: Typedef("__jmp_buf_tag"),
    //         taudt_bits: SDACL(TypeAttribute(0))
    //       })
    //     }),
    //     None
    //   )],
    //   retloc: None }
    // ```
    let function = [
        0x0c, // func type
        0x13, // TODO
        0x07, // return int
        0x02, // 1 parameter
        0xff, 0x48, // TODO
        0x0a, // arg1 type pointer
        0xfe, 0x10, // TypeAttribute val
        0x02, // dt len 1
        0x0d, 0x5f, 0x5f, 0x6f, 0x72, 0x67, 0x5f, 0x61, 0x72, 0x72, 0x64, 0x69,
        0x6d, // TODO some _string: "__org_arrdim"
        0x03, 0xac, 0x01, // TODO _other_thing
        0x0d, // arg1 pointer type struct
        0x01, // struct ref
        0x0e, 0x5f, 0x5f, 0x6a, 0x6d, 0x70, 0x5f, 0x62, 0x75, 0x66, 0x5f, 0x74,
        0x61, 0x67, // "__jmp_buf_tag"
        0x00, // end of type
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_aes_encrypt() {
    // ```c
    // void AES_ctr128_encrypt(
    //   const unsigned __int8 *in,
    //   unsigned __int8 *out,
    //   const unsigned int length,
    //   const AES_KEY *key,
    //   unsigned __int8 ivec[16],
    //   unsigned __int8 ecount_buf[16],
    //   unsigned int *num
    // );
    // ```
    let function = [
        0x0c, // type function
        0x13, 0x01, // ???
        0x08, // 7 args
        // arg1 ...
        0x0a, // pointer
        0x62, // const unsigned __int8
        // arg2 ...
        0x0a, // pointer
        0x22, // unsigned __int8
        // arg3 ...
        0x67, // const unsigned int
        // arg4 ...
        0x0a, // pointer
        0x7d, // const typedef
        0x08, 0x41, 0x45, 0x53, 0x5f, 0x4b, 0x45,
        0x59, // ordinal "AES_KEY"
        // arg5
        0xff, 0x48, // some flag in function arg
        0x0a, // pointer
        0xfe, 0x10, // TypeAttribute val
        0x02, // TypeAttribute loop once
        0x0d, 0x5f, 0x5f, 0x6f, 0x72, 0x67, 0x5f, 0x61, 0x72, 0x72, 0x64, 0x69,
        0x6d, // string "__org_arrdim"
        0x03, 0xac, 0x10, // ???? some other TypeAttribute field
        0x22, // type unsigned __int8
        // arg6
        0xff, 0x48, // some flag in function arg
        0x0a, // pointer
        0xfe, 0x10, // TypeAttribute val
        0x02, // TypeAttribute loop once
        0x0d, 0x5f, 0x5f, 0x6f, 0x72, 0x67, 0x5f, 0x61, 0x72, 0x72, 0x64, 0x69,
        0x6d, // string "__org_arrdim"
        0x03, 0xac, 0x10, // ???? some other TypeAttribute field
        0x22, // type unsigned __int8
        // arg7 ...
        0x0a, // pointer
        0x27, // unsigned int
        0x00,
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_spoiled_function_kernel_32() {
    // ```
    // TilType(Type { is_const: false, is_volatile: false, type_variant:
    //   Function(Function {
    //     ret: Type { is_const: false, is_volatile: false, type_variant: Basic(Void) },
    //     args: [
    //       (
    //         Some([117, 69, 120, 105, 116, 67, 111, 100, 101]),
    //         Type {
    //           is_const: false,
    //           is_volatile: false,
    //           type_variant: Typedef(Name([85, 73, 78, 84])) }, None)],
    //           retloc: None
    //         }
    //       )
    // })
    // ```
    let function = [
        0x0c, // function type
        0xaf, 0x81, // function cc extended...
        0x42, // flag
        0x01, // 0 regs nspoiled
        0x53, // cc
        0x01, // return type void
        0x02, // 1 param
        0x3d, // param 1 typedef
        0x05, 0x55, 0x49, 0x4e, 0x54, // typedef name
        0x00, //end
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_spoiled_function_invalid_reg() {
    // ```
    // 0x180001030:
    // TilType(Type { is_const: false, is_volatile: false, type_variant:
    //   Function(Function {
    //     ret: Type { is_const: false, is_volatile: false, type_variant: Basic(Void) },
    //     args: [], retloc: None
    //   })
    // })
    // ```
    let function = [
        0x0c, // function type
        0xaa, // extended function cc, 10 nspoiled
        0x71, // spoiled reg 0
        0x72, // spoiled reg 1
        0x73, // spoiled reg 2
        0x79, // spoiled reg 3
        0x7a, // spoiled reg 4
        0x7b, // spoiled reg 5
        0x7c, // spoiled reg 6
        0xc0, 0x08, // spoiled reg 7
        0xc4, 0x08, // spoiled reg 8
        0xc5, 0x08, // spoiled reg 9
        0x30, // cc
        0x01, // return type void
        0x01, // no params
        0x00, // end
    ];
    let _til = til::Type::new_from_id0(&function, vec![]).unwrap();
}

#[test]
fn parse_struct_with_fixed() {
    let function = [
        0x0d, // stuct type
        0x31, // n = 0x30, mem_cnt = 6, packalig = 0
        0xf1, 0x80, 0x08, // struct att
        0x32, // member 0 => char
        0x01, // member 0 fixed_ext_att
        0x03, // member 1 => int16
        0x02, 0x10, // member 1 fixed_ext_att
        0x07, // member 2 => int
        0x02, 0x10, // member 2 fixed_ext_att
        0x3d, 0x03, 0x23, 0x48, // member 3 => typeref(8)
        0x02, 0x20, // member 3 fixed_ext_att
        0x08, // member 4 => bool
        0x02, 0x40, // member 4 fixed_ext_att
        0x1b, // member 5 array
        0x01, // member 5 nelem = 0
        0x32, // member 5 inner_type = char
        0x02, 0x08, // member 5 fixed_ext_att
        0x02, 0x13, // struct stuff
        0x00, //end
    ];
    let til = til::Type::new_from_id0(&function, vec![]).unwrap();
    let til::TypeVariant::Struct(til_struct) = til.type_variant else {
        unreachable!()
    };
    assert!(til_struct.extra_padding == Some(19));
}

#[rstest]
fn parse_til(
    #[files("resources/tils/**/*.til")]
    #[exclude("resources/tils/local/.*")]
    file: PathBuf,
) {
    parse_til_inner(file)
}

#[rstest]
fn parse_local_til(#[files("resources/tils/local/**/*.til")] file: PathBuf) {
    parse_til_inner(file)
}

#[rstest]
fn parse_idb(
    #[files("resources/idbs/**/*.i64")]
    #[files("resources/idbs/**/*.idb")]
    #[exclude("resources/idbs/local/.*")]
    filename: PathBuf,
) {
    parse_idb_inner(filename)
}

#[rstest]
fn parse_local_idb(
    #[files("resources/idbs/local/**/*.i64")]
    #[files("resources/idbs/local/**/*.idb")]
    filename: PathBuf,
) {
    parse_idb_inner(filename)
}

fn remove_base_dir<'a>(
    file: &'a Path,
    resource: &'static str,
) -> impl Iterator<Item = std::path::Component<'a>> {
    let root_env = env!("CARGO_MANIFEST_DIR");
    let root = Path::new(root_env).join("resources").join(resource);
    if !file.starts_with(&root) {
        panic!("Invalid Path {root_env:?}: {:?}", file.to_str());
    }
    let len = root.components().count();
    file.components().skip(len)
}

fn parse_til_inner(file: PathBuf) {
    let file_suffix: PathBuf = remove_base_dir(&file, "tils").collect();
    let file_suffix_str = file_suffix.to_str().unwrap();
    println!("{file_suffix_str}");
    // makes sure it don't read out-of-bounds
    let mut input = BufReader::new(File::open(file).unwrap());
    // TODO make a SmartReader
    let til = TILSection::read(&mut input).unwrap();
    insta::with_settings!({snapshot_suffix => file_suffix_str, sort_maps => true}, {
        insta::assert_yaml_snapshot!(til);
    });

    assert_eq!(
        input.peek_u8().unwrap(),
        None,
        "unable to consume the entire TIL file"
    );
}

fn parse_idb_inner(file: PathBuf) {
    let file_suffix: PathBuf = remove_base_dir(&file, "idbs").collect();
    println!("{}", file_suffix.to_str().unwrap());
    let mut input = BufReader::new(File::open(&file).unwrap());
    let format = identify_idb_file(&mut input).unwrap();
    insta::with_settings!({sort_maps => true}, {
        match format {
            IDBFormats::Separated(IDAVariants::IDA32(sections)) => {
                parse_idb_separated(file_suffix, &mut input, &sections)
            }
            IDBFormats::Separated(IDAVariants::IDA64(sections)) => {
                parse_idb_separated(file_suffix, &mut input, &sections)
            }
            IDBFormats::InlineUncompressed(sections) => {
                parse_idb_inlined(file_suffix, &mut input, &sections)
            }
            IDBFormats::InlineCompressed(compressed) => {
                let mut decompressed = Vec::new();
                let sections = compressed
                    .decompress_into_memory(input, &mut decompressed)
                    .unwrap();
                parse_idb_inlined(
                    file_suffix,
                    &mut Cursor::new(decompressed),
                    &sections,
                );
            }
        }
    });
}

fn parse_idb_separated<K: IDAKind, I: BufRead + Seek>(
    file_suffix: PathBuf,
    input: &mut I,
    sections: &SeparatedSections<K>,
) {
    // parse sectors
    let id0 = sections
        .read_id0(&mut *input, sections.id0_location().unwrap())
        .unwrap();
    let id1 = sections
        .read_id1(&mut *input, sections.id1_location().unwrap())
        .unwrap();
    let id2 = sections
        .id2_location()
        .map(|id2| sections.read_id2(&mut *input, id2))
        .transpose()
        .unwrap();
    let til = sections
        .til_location()
        .map(|til| sections.read_til(&mut *input, til).unwrap());
    insta::with_settings!({snapshot_suffix => file_suffix.join("til").to_str().unwrap()}, {
        insta::assert_yaml_snapshot!(til);
    });
    let _nam = sections
        .nam_location()
        .map(|idx| sections.read_nam(&mut *input, idx));
    parse_idb_data(file_suffix, &id0, &id1, id2.as_ref(), til.as_ref())
}

fn parse_idb_inlined<I: BufRead + Seek>(
    file_suffix: PathBuf,
    input: &mut I,
    sections: &InlineUnCompressedSections,
) {
    // parse sectors
    let id0 = sections
        .read_id0(&mut *input, sections.id0_location().unwrap())
        .unwrap();
    let id1 = sections
        .read_id1(&mut *input, sections.id1_location().unwrap())
        .unwrap();
    let id2 = sections
        .id2_location()
        .map(|id2| sections.read_id2(&mut *input, id2))
        .transpose()
        .unwrap();
    let til = sections
        .til_location()
        .map(|til| sections.read_til(&mut *input, til).unwrap());
    let _nam = sections
        .nam_location()
        .map(|idx| sections.read_nam(&mut *input, idx));
    parse_idb_data(file_suffix, &id0, &id1, id2.as_ref(), til.as_ref())
}

fn parse_idb_data<K>(
    file_suffix: PathBuf,
    id0: &ID0Section<K>,
    id1: &ID1Section<K>,
    id2: Option<&ID2Section<K>>,
    til: Option<&TILSection>,
) where
    K: IDAKind,
{
    // parse all id0 information
    let root_netnode = id0.root_node().unwrap();
    let ida_info = id0.ida_info(root_netnode.into()).unwrap();
    insta::with_settings!({snapshot_suffix => file_suffix.join("ida_info").to_str().unwrap()}, {
        insta::assert_yaml_snapshot!(ida_info);
    });

    insta::with_settings!({snapshot_suffix => file_suffix.join("segments").to_str().unwrap()}, {
        let seg_idx = id0.segments_idx().unwrap().unwrap();
        let mut segments: Vec<Segment<K>> =
            id0.segments(seg_idx).map(Result::unwrap).collect();
        segments.sort_unstable_by_key(|seg| (seg.address.start, seg.address.end, seg.selector));
        insta::assert_yaml_snapshot!(segments);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("loader_name").to_str().unwrap()}, {
        let loader_name: Option<Vec<&str>> = id0
            .loader_name()
            .unwrap()
            .map(|iter| iter.map(Result::unwrap).collect());
        insta::assert_yaml_snapshot!(loader_name);
    });
    let root_info_idx = id0.root_node().unwrap();
    // I belive the input file should always be present, but maybe I'm wrong,
    // I need know if this unwrap panics
    insta::with_settings!({snapshot_suffix => file_suffix.join("input_file").to_str().unwrap()}, {
        let input_file = id0.input_file(root_info_idx).unwrap();
        insta::assert_yaml_snapshot!(input_file);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("input_file_size").to_str().unwrap()}, {
        let input_file_size = id0.input_file_size(root_info_idx).unwrap();
        insta::assert_yaml_snapshot!(input_file_size);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("input_file_crc32").to_str().unwrap()}, {
        let input_file_crc32 =
            id0.input_file_crc32(root_info_idx).unwrap().unwrap();
        insta::assert_yaml_snapshot!(input_file_crc32);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("input_file_sha256").to_str().unwrap()}, {
        let input_file_sha256 = id0.input_file_sha256(root_info_idx).unwrap();
        insta::assert_yaml_snapshot!(input_file_sha256);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("input_file_md5").to_str().unwrap()}, {
        let input_file_md5 = id0.input_file_md5(root_info_idx).unwrap();
        insta::assert_yaml_snapshot!(input_file_md5);
    });
    // TODO I think database information is always available, check that...
    insta::with_settings!({snapshot_suffix => file_suffix.join("database_num_opens").to_str().unwrap()}, {
        let database_num_opens =
            id0.database_num_opens(root_info_idx).unwrap().unwrap();
        insta::assert_yaml_snapshot!(database_num_opens);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("database_secs_opens").to_str().unwrap()}, {
        let database_secs_opens =
            id0.database_secs_opens(root_info_idx).unwrap().unwrap();
        insta::assert_yaml_snapshot!(database_secs_opens);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("database_creation_time").to_str().unwrap()}, {
        let database_creation_time =
            id0.database_creation_time(root_info_idx).unwrap().unwrap();
        insta::assert_yaml_snapshot!(database_creation_time);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("database_initial_version").to_str().unwrap()}, {
        let database_initial_version = id0
            .database_initial_version(root_info_idx)
            .unwrap()
            .unwrap();
        insta::assert_yaml_snapshot!(database_initial_version);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("database_creation_version").to_str().unwrap()}, {
        let database_creation_version =
            id0.database_creation_version(root_info_idx);
        insta::assert_yaml_snapshot!(database_creation_version);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("c_predefined_macros").to_str().unwrap()}, {
        let c_predefined_macros = id0.c_predefined_macros(root_info_idx);
        insta::assert_yaml_snapshot!(c_predefined_macros);
    });
    insta::with_settings!({snapshot_suffix => file_suffix.join("c_header_path").to_str().unwrap()}, {
        let c_header_path = id0.c_header_path(root_info_idx);
        insta::assert_yaml_snapshot!(c_header_path);
    });
    // TODO identify the data
    //let Some(_) = id0.output_file_encoding_idx(root_info_idx) else {todo!()};
    //let Some(_) = id0.ids_modenode_id(root_info_idx) else {todo!()};
    //id0.user_closed_source_files(root_info_idx).unwrap().collect();
    //let Some(_) = id0.problem_lists(root_info_idx) else {todo!()};
    //let Some(_) = id0.archive_file_path(root_info_idx) else {todo!()};
    //let Some(_) = id0.abi_name(root_info_idx) else {todo!()};
    //id0.debug_binary_paths(root_info_idx).unwrap().collect();
    //let Some(_) = id0.strings_encodings(root_info_idx) else {todo!()};
    //let Some(_) = id0.text_representation_options(root_info_idx) else {todo!()};
    //let Some(_) = id0.graph_representation_options(root_info_idx) else {todo!()};
    //let Some(_) = id0.instant_idc_statements(root_info_idx) else {todo!()};
    //let Some(_) = id0.assembler_include_filename(root_info_idx) else {todo!()};
    //id0.notepad_data(root_info_idx).unwrap().collect();
    //let Some(_) = id0.instant_idc_statements_old(root_info_idx) else {todo!()};
    //let Some(_) = id0.segment_group_info(root_info_idx) else {todo!()};
    //id0.selectors(root_info_idx).unwrap().collect();
    //let Some(_) = id0.file_format_name_loader(root_info_idx) else {todo!()};

    insta::with_settings!({snapshot_suffix => file_suffix.join("file_regions").to_str().unwrap()}, {
        let file_regions_idx = id0.file_regions_idx().unwrap();
        let file_regions: Vec<FileRegions<K>> = id0
            .file_regions(file_regions_idx, ida_info.version)
            .map(Result::unwrap)
            .collect();
        insta::assert_yaml_snapshot!(file_regions);
    });
    if let Some(func_idx) = id0.funcs_idx().unwrap() {
        let _functions_and_comments: Vec<FunctionsAndComments<'_, K>> = id0
            .functions_and_comments(func_idx)
            .map(Result::unwrap)
            .collect();
    }
    insta::with_settings!({snapshot_suffix => file_suffix.join("entry_points").to_str().unwrap()}, {
        let entry_points = id0.entry_points().unwrap();
        insta::assert_yaml_snapshot!(entry_points);
    });
    let _ = id0.dirtree_bpts().unwrap();
    let _ = id0.dirtree_enums().unwrap();

    if let Some(dirtree_names) = id0.dirtree_names().unwrap() {
        let image_base = ida_info.netdelta();
        dirtree_names.visit_leafs(|addr| {
            // NOTE it's know that some labels are missing from the byte
            // info but not from the databases, maybe in cases they are
            // created in debug-memory-pages or similar...
            let addr_info = crate::addr_info::AddressInfo::new(
                id0,
                id1,
                id2,
                image_base,
                Address::from_raw(*addr),
            )
            .or_else(|| {
                // TODO make sure this new_forced is required
                crate::addr_info::AddressInfo::new_forced(
                    id0,
                    image_base,
                    Address::from_raw(*addr),
                )
            })
            .unwrap();
            let _name = addr_info.label().unwrap();
        });
    }
    if let Some((_dirtree_tinfos, til)) = id0.dirtree_tinfos().unwrap().zip(til)
    {
        _dirtree_tinfos.visit_leafs(|ord| {
            let _til = til.get_ord((*ord).into()).unwrap();
        });
    }
    let _ = id0.dirtree_imports().unwrap();
    let _ = id0.dirtree_structs().unwrap();
    let _ = id0.dirtree_function_address().unwrap();
    let _ = id0.dirtree_bookmarks_tiplace().unwrap();
    let _ = id0.dirtree_bookmarks_idaplace().unwrap();
    let _ = id0.dirtree_bookmarks_structplace().unwrap();
}
