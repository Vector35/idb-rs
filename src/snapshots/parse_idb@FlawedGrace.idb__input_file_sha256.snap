---
source: src/test.rs
expression: input_file_sha256
---
- 176
- 61
- 10
- 167
- 163
- 23
- 56
- 182
- 138
- 10
- 231
- 216
- 21
- 57
- 47
- 42
- 217
- 141
- 233
- 78
- 35
- 204
- 124
- 235
- 135
- 32
- 99
- 99
- 186
- 239
- 176
- 185
